Generators semantics

jugglinmike opened the pull request:
@bterlson I'm submitting this as two commits to separate the tests that are derived directly from V8 from the analogous tests I added to fill out the coverage

arv commented on test/built-ins/GeneratorPrototype/next/incorrect-context.js:
```diff
@@ -0,0 +1,22 @@
+function* g() {}
+var iter = g();
+
+assert.throws(TypeError, function() { iter.next.call(1); });
```
Would it be better to test this with `%IteratorPrototype%.next`?

arv replied:
Sorry, `%GeneratorPrototype%.

jugglinmike replied:
That would simplify things a bit (no implicit prototype resolution and no need to create an iterable). It would introduce a dependency on `Object.getPrototypeOf`, but that's probably still better overall. I'm pushing up a commit to address this now!

arv commented on test/built-ins/GeneratorPrototype/throw/incorrect-context.js:
```diff
@@ -0,0 +1,17 @@
+function* g() {}
+var iter = g();
+
+assert.throws(TypeError, function() { iter.throw.call(1); });
```
same here?

arv commented on test/built-ins/GeneratorPrototype/throw/try-catch-before-try.js:
```diff
@@ -0,0 +1,35 @@
+result = iter.next();
+assert.sameValue(result.value, 1, 'First result `value`');
+assert.sameValue(result.done, false, 'First result `done` flag');
+assert.throws(Test262Error, function() { iter.throw(new Test262Error()); });
```
Use a different constructor function here

arv commented on test/built-ins/GeneratorPrototype/throw/try-catch-following-catch.js:
```diff
@@ -0,0 +1,43 @@
+}
+var iter, result, exception;
+
+exception = new Test262Error();
```
It seems bad to use Test262Error for these kind of tests.

bterlson replied:
The cat is already out of the bag. The python harness (somehow) always injects a definition of Test262Error into the test environment, and many of the recent tests use it. I just had to update test262-harness to work with this yesterday :)

arv replied:
My concern is that there might be some false positives (or negatives) when using these but maybe I'm worrying too much?

bterlson replied:
I can't think of a case where you'd get false behavior. It's equivalent in risk to making your own error constructor I think.

arv commented on test/language/expressions/generators/length-property-descriptor.js:
```diff
@@ -0,0 +1,15 @@
+
+verifyNotEnumerable(g, 'length');
+verifyNotWritable(g, 'length');
+verifyConfigurable(g, 'length');
```
Verify that the length is 0?

jugglinmike replied:
Unlike `name`, there aren't any tests for the value of the length property. I agree it's worth adding a test now, but I'll wait on Brian's response to https://github.com/tc39/test262/pull/237#discussion_r29007374 before acting on this (since I want to include the new assertions in a way that is consistent with the others).

arv commented on test/language/expressions/generators/name-property-descriptor.js:
```diff
@@ -0,0 +1,15 @@
+
+verifyNotEnumerable(g, 'name');
+verifyNotWritable(g, 'name');
+verifyConfigurable(g, 'name');
```
Verify that the name is 'g'?

jugglinmike replied:
There's a separate test for the value of the `name` property in `master` already: https://github.com/tc39/test262/blob/c49e2bd6b902aa699bedeabcf6e84adf99688a99/test/language/expressions/generators/name.js

@bterlson Do you think it's worthwhile to keep these assertions separate? Or should I merge the existing file into this one?

bterlson replied:
Seems ok to be separate. But generally, a descriptor contains a value property, so it makes some sense to combine them - you're testing all the descriptor properties.

arv commented on test/language/statements/generators/length-property-descriptor.js:
```diff
@@ -0,0 +1,15 @@
+
+verifyNotEnumerable(g, 'length');
+verifyNotWritable(g, 'length');
+verifyConfigurable(g, 'length');
```
same

arv commented on test/language/statements/generators/name-property-descriptor.js:
```diff
@@ -0,0 +1,15 @@
+
+verifyNotEnumerable(g, 'name');
+verifyNotWritable(g, 'name');
+verifyConfigurable(g, 'name');
```
same

jugglinmike commented:
I've addressed all your feedback, @arv. Thanks for the review!

anba commented on test/built-ins/GeneratorPrototype/next/context-constructor-invocation.js:
```diff
@@ -0,0 +1,16 @@
+---*/
+
+var context;
+function* g() { context = this; }
```
Access to `this` in new'ed generators throws a ReferenceError. (Generators are created as derived constructor functions.)

jugglinmike replied:
This caught me off guard. Doing my due diligence, I verified in the spec:

> **9.2.3 FunctionAllocate (functionPrototype, strict [,functionKind] )**
> ...
> 9. If needsConstruct is true, then
>    a. Set F’s [[Construct]] internal method to the definition specified in
>       9.2.2.
>    b. If functionKind is "generator", set the [[ConstructorKind]] internal
>       slot of F to "derived".
>    c. Else, set the [[ConstructorKind]] internal slot of F to "base".
>    d. NOTE Generator functions are tagged as "derived" constructors to
>       prevent [[Construct]] from preallocating a generator instance.
>       Generator instance objects are allocated when EvaluateBody is applied
>       to the GeneratorBody of a generator function.

and:

> **9.2.2 `[[Construct]]` ( argumentsList, newTarget)**
> ...
> 5. If kind is "base", then
>    a. Let thisArgument be OrdinaryCreateFromConstructor(newTarget,
>       "%ObjectPrototype%").
>    b. ReturnIfAbrupt(thisArgument).

Thanks!

anba commented on test/language/expressions/yield/star-iterable.js:
```diff
@@ -0,0 +1,43 @@
+    should produce an iterator that visits each iterated item.
+---*/
+
+var results = [{ value: 1 }, 13, 'foo', { value: 34, done: true }];
```
Non-object types throw a TypeError, cf. 7.4.2 IteratorNext step 4.

jugglinmike replied:
Right you are. I think that behavior warrants a dedicated test, so I think I will simplify this to return a valid object for each iteration.

anba commented on test/language/expressions/generators/name-property-descriptor.js:
```diff
@@ -0,0 +1,16 @@
+
+var g = function*() {};
+
+assert.sameValue(g.name, '');
```
`g.name` is "g".

jugglinmike commented:
@bterlson I've squashed this patch into two distinct commits to maintain the distinction between the tests inspired by the V8 test suite directly, and the tests I added to fill out coverage a bit.
